rt/frontend/jaxrs/src/main/java/org/apache/cxf/jaxrs/model/URITemplate.java

```diff
@@ -130,10 +130,31 @@ public String getPatternValue() {
 
     private static String escapeCharacters(String expression) {
 
-        StringBuilder sb = new StringBuilder();
-        for (int i = 0; i < expression.length(); i++) {
-            char ch = expression.charAt(i);
-            sb.append(isReservedCharacter(ch) ? "\\" + ch : ch);
+        int length = expression.length();
```

Comment:
My apologies but this implementation is not very convincing to me, I think we could just pre-allocate `2 * expression.length()` using `StringBuilder ` constructor (worst case when all characters are reserved) and keep the previous one?

Reply:
The goal is to reduce the number of StringBuilders created in order to reduce GCs and CPU usage. In this case we don't create a StringBuilder until we are sure that we actually need one, vs creating one that we might not even use.

Reply:
Thanks, from this perspective we could consider it as micro-optimization, trading the code readability, I think it is acceptable in this case, the implementation could be understood easily.

Reply:
This being micro-optimization or not should really be supported by benchmark result data. If this is on a hot path, creating less StringBuilder can really make some differences.
This said, please note that the fix actually solves 2 issue:
1) it avoids creating the StringBuilder when not needed
2) it prevents useless StringBuilder instances being created whenever a reserved character is found, as sb.append(isReservedCharacter(ch) ? "\\" + ch : ch) internally creates another StringBuilder to perform the "\\" + ch concatenation.

Reply:
We don't have hard numbers at the moment, but all of these changes combined are estimated to improve performance by 1.5%-2% per request.

Our JDK team analyzed our performance in order to reduce CPU usage and garbage collection and found that we are creating a large number of StringBuilders. StringBuilders are expensive to use so only using them when we need them will improve performance. Similarly, they found that we were using reflection to get the params every request in ResourceUtils.createConstructorArguments() which is also expensive, hence the change to get the params one time in the constructor.